### PR TITLE
Adding TLC stats to cimetrics

### PR DESCRIPTION
**Pull request by heidihoward**
This PR adds some TLC stats to the [cimetrics](https://github.com/jumaffre/cimetrics). To keep things simple, this PR just adds the duration and number of distinct states for `MCccfraft.tla` but we can easily extended this.

### CODE REVIEW

**Comment by heidihoward** [comment]
@eddyashton @achamayou Still a WIP but I'd appreciate feedback at this stage. I've deliberately omitted `METRICS_MONGO_CONNECTION` from the gh action to avoid actually posting data until we are ready

File: tests/upload_tlc_stats.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        with open(filename) as f:
+            data = json.load(f)
+            duration_sec = data["duration"]
+            dstates = data["distinct"]
+            LOG.info(
+                "Uploading metrics - duration: {}, distinct states: {}",
+                duration_sec,
+                dstates,
+            )
+            with cimetrics.upload.metrics(complete=False) as metrics:
```
Comment (lemmy) [inline tests/upload_tlc_stats.py:23]: I don't know what cimetrics provides but https://github.com/tlaplus/Examples/blob/62d104f2e1cfd63b3f928342b8f634298706d145/specifications/ewd998/EWD998ChanID_export.tla#L46-L56 is an example about how you can do post requests from within TLC, which would make the json serialization obsolete.

Comment (heidihoward) [reply tests/upload_tlc_stats.py:23]: This is cimetric https://github.com/jumaffre/cimetrics, it provides the fancy performance graphs in each PR. 

I think the Json stats file might be useful in the future to quickly compare the performance of different configurations

Comment (lemmy) [reply tests/upload_tlc_stats.py:23]: If more is going on in cimetrics than just a basic http post, the .py script is the way to go. :-)

Btw. I do have the CosmosDB client library wrapped as a TLA+ module somewhere, but that's just an interesting hack.

Comment (achamayou) [reply tests/upload_tlc_stats.py:23]: Upload is pretty straightforward: https://github.com/jumaffre/cimetrics/blob/main/cimetrics/upload.py, most of the logic is there to the querying and the fancy plotting (which you can see in this PR too if you scroll up a bit).

For procedural tasks such as this one, Python has a certain charm :)

**Review by achamayou — APPROVED** [review]

**Comment by achamayou** [comment]
@heidihoward if you would like a different category, you can add a line in https://github.com/microsoft/CCF/blob/main/metrics.yml, something like "TLC": "tlc_.*", before the "Others" catch-all (they are evaluated in order).

**Comment by heidihoward** [comment]
> @heidihoward if you would like a different category, you can add a line in https://github.com/microsoft/CCF/blob/main/metrics.yml, something like "TLC": "tlc_.*", before the "Others" catch-all (they are evaluated in order).

done in https://github.com/microsoft/CCF/pull/5807/commits/e1e3cc4acd46d510acc432c2efd4507b89bde4ff

File: .azure-pipelines-templates/model_checking.yml [hunk 2]
```diff
@@ -23,13 +23,21 @@ jobs:
...
 
-      - script: ./tlc.sh -workers auto -config consensus/MCccfraftWithReconfig.cfg consensus/MCccfraft.tla -dumpTrace tla MCccfraftWithReconfig.trace.tla -dumpTrace json MCccfraftWithReconfig.json
-        workingDirectory: tla
-        displayName: MCccfraftWithReconfig.cfg
+      # - script: ./tlc.sh -workers auto -config consensus/MCccfraftWithReconfig.cfg consensus/MCccfraft.tla -dumpTrace tla MCccfraftWithReconfig.trace.tla -dumpTrace json MCccfraftWithReconfig.json
+      #   workingDirectory: tla
+      #   displayName: MCccfraftWithReconfig.cfg
+
+      - script: |
+          pip install -r requirements.txt
```
Comment (achamayou) [inline .azure-pipelines-templates/model_checking.yml:35]:
```suggestion
          set -ex
          python3.8 -m venv env
          source env/bin/activate
          python -m pip install -r requirements.txt
```

File: .azure-pipelines-templates/model_checking.yml [hunk 3]
```diff
@@ -23,13 +23,24 @@ jobs:
...
+      # - script: ./tlc.sh -workers auto -config consensus/MCccfraftWithReconfig.cfg consensus/MCccfraft.tla -dumpTrace tla MCccfraftWithReconfig.trace.tla -dumpTrace json MCccfraftWithReconfig.json
+      #   workingDirectory: tla
+      #   displayName: MCccfraftWithReconfig.cfg
+
+      - script: |
+          set -ex
+          python3.8 -m venv env
+          source env/bin/activate
+          python -m pip install -r requirements.txt
+          python3 upload_tlc_stats.py 3node_fixed ../tla/MCccfraft.json
```
Comment (achamayou) [inline .azure-pipelines-templates/model_checking.yml:39]:
```suggestion
          ls ../tla/*.json
          python3 upload_tlc_stats.py 3node_fixed ../tla/MCccfraft_stats.json
```

**Comment by heidihoward** [comment]
Yay, this is now working! 
image

**Review by achamayou — APPROVED** [review]

**Comment by heidihoward** [comment]
Looking at the CI output wheel is complaining a bit but it seems to work anyways. I think actually it might be best to switch to [cimetrics_env.sh](https://github.com/microsoft/CCF/blob/main/tests/cimetrics_env.sh) so only cimetrics is installed. I did spot this handy script the first time around but did not use it as I wrongly assumed we didn't need a virtual env.

PS: Sorry, I don't know how to post a link to ADO pipelines, copying and pasting the URL gives this fun error 
image

**Comment by achamayou** [comment]
@heidihoward you're absolutely right, I'd completely forgotten about that script.

**Pull request merged**